feat: IDirectoryInfo.CopyTo

mayermart opened the pull request:
Hi,

i implemented a first draft for IDirectoryInfo.CopyTo for recursively copying an IDirecotyInfo to a new path.
There are two possible usages:
`IDirectoryInfo IDirectoryInfo.CopyTo(string destDirectoryName, bool recursive)` and
`IDirectoryInfo IDirectoryInfo.CopyTo(IDirectoryInfo destDirectory, bool recursive)`

Please take a look and provide some feedback :-)

This closes #18

gigi81 commented on src/System.IO.Abstractions.Extensions/IDirectoryInfoExtensions.cs:
```diff
@@ -3,25 +3,103 @@
+            }
+
+            destDirectory.Refresh();
+            if (destDirectory.Exists)
```
Why throw if the destination directory exists? I don't think this check is needed. can you remove it?

gigi81 commented on src/System.IO.Abstractions.Extensions/IDirectoryInfoExtensions.cs:
```diff
@@ -3,25 +3,103 @@
+            }
+
+            var fileSystem = info.FileSystem;
+            fileSystem.Directory.CreateDirectory(destDirectory.FullName);
```
I might be wrong but I think a create method exists like destDirectory.Create() or not?

mayermart replied:
I thought destDirectory.Create() would not create the hierarchy if a parent does not exist. With destDirectory.Create() it looks much nicer.

gigi81 commented on src/System.IO.Abstractions.Extensions/IDirectoryInfoExtensions.cs:
```diff
@@ -3,25 +3,103 @@
+
+            foreach (var fileInfo in info.EnumerateFiles())
+            {
+                string targetFilePath = Path.Combine(destDirectory.FullName, fileInfo.Name);
```
Path.Combine is not the right method. use the equivalent in filesystem

gigi81 commented:
hi @mayermart  thanks for your PR. please have a look at my comments

gigi81 commented on src/System.IO.Abstractions.Extensions/IDirectoryInfoExtensions.cs:
```diff
@@ -3,25 +3,103 @@
     public static class IDirectoryInfoExtensions
     {
         /// <summary>
-        /// Get an <see cref="IDirectoryInfo"/> for the specified sub-directory <paramref name="name"/>
```
Also can you please move the new methods at the bottom of the file instead of the top so the merge looks cleaner

mayermart commented:
Hi @gigi81, thanks for your feedback.
I implemented all your recommendations and reformatted the files for cleaner merge.

Actually i can't remember why i implemented the throw if the destination already exists. Its not even in the template from MS.

gigi81 commented:
Thanks! Will review later!

gigi81 commented:
Merging this into a feature branch as there are still some bits I'm not totally convinced and I want to fix before merging into main. Thanks for your contribution
